Minor mapper enhancements

ulupo opened the pull request:
#### What does this implement/fix? Explain your changes.
- Changes the default parameter for `node_color_statistic` in `plot_static_mapper_graph` and `plot_interactive_mapper_graph` from `np.mean` to `None`. This is consistent with the treatment of defaults elsewhere in giotto-tda and makes for better-looking documentation, avoiding such horrors as `node_color_statistic=<function mean at 0x10d7d6ef0>`.
- Wording changes and docstring improvements
- Add new `max_fraction` parameter to agglomerative clusterers to address, addressing a variant of (#130)
- Add hyperparameter validation to cover classes and flat-cut clusterers
- Print closed instead of open intervals in `utils.validation.validate_params`

Commit:
Change default parameter for node_color_statistic to None

This is consistent with the treatment of defaults elsewhere in giotto-tda and makes for better-looking documentation.

Commit:
Fix "is None" -> "is not None"

Commit:
Wording changes in mapper_quickstart.ipynb

lewtun commented on gtda/mapper/visualization.py:
```diff
@@ -94,13 +95,18 @@ def plot_static_mapper_graph(
         pipe = pipeline
 
+    if node_color_statistic is not None:
+        _node_color_statistic = node_color_statistic
```
I wonder whether this distinction between protected and unprotected variables is still needed? Could we get away with just using `node_color_statistic`?

ulupo replied:
I've always seen this also as a way to guide the developer or the person reading the code in general. Technically, I think the code would work even without it in this case (as ``None`` is immutable).

lewtun replied:
Fine with me - let's keep it

lewtun reviewed (commented):
Will give approval once the final PR is ready for review

Commit:
Add literature reference for FirstHistogramGap

Commit:
Add URL for future Mapper pipeline diagram

Commit:
Add max_fraction parameter to agglomerative clusterers, add validation

- max_fraction can be used to limit the number of clusters found by FirstSimpleGap and FirstHistogramGap
- Implement hyperparameter validation via gtda.utils.validation in agglomerative clusterers

Commit:
Systematically Implement hyperparameter validation in mapper covers

- Implement hyperparameter validation via gtda.utils.validation in covering transformers
- Remove superseded _validate_kind utility function

Commit:
Open -> closed intervals in ValueError messages from validate_params

lewtun commented on gtda/mapper/cluster.py:
```diff
@@ -322,14 +324,23 @@ class FirstSimpleGap(ClusterMixin, BaseEstimator, Agglomerative):
     first occurrence of a gap, between the linkage values of successive merges,
-    which exceeds this threshold. Otherwise, a single cluster is returned.
+    which exceeds this threshold. Otherwise, a single cluster is returned. The
+    algorithm can partially overridden to ensure that the final number of
```
"can be"

ulupo replied:
Done.

lewtun commented on gtda/mapper/cluster.py:
```diff
@@ -433,21 +454,29 @@ def fit(self, X, y=None):
     bin of height no greater than f (i.e. a "gap") is observed; 3) if no gap is
-    observed, increase k and repeat 1) and 2) until termination.
+    observed, increase k and repeat 1) and 2) until termination. The algorithm
+    can partially overridden to ensure that the final number of clusters does
```
"can be"

ulupo replied:
Done.

lewtun commented on gtda/mapper/pipeline.py:
```diff
@@ -154,14 +154,15 @@ def make_mapper_pipeline(scaler=None,
+    or a callable acting on one-dimensional arrays -- in the latter case,
+    a transformer is internally created whose ``fit_transform`` applies this
+    callable independently on each row of the data. The clustering step need
+    only implement a ``fit`` method storing clustering labels.
```
"needs to only implement a ``fit`` method that stores the clustering labels."

Commit:
Fix mistakes in docstrings concerning max_fraction

Commit:
Fix incorrect definition of _hyperparameters in coverers

Commit:
Fix _hyperparameters 'n_bin_start' key in agglo clusterers, fix types

Commit:
Fix forgotten refactored variable

Commit:
Use ceil instead of np.ceil for correct type

Commit:
Use int instead of numbers.Number in freq_threshold validation
